[Non_Sanity_Testing] Test case "test_paging_with_mme_restart.py" is flaky

### Your Environment

- **Version:** Local Development Setup
- **Affected Component:** Access Gateway
- **Affected Subcomponent:** Magma MME/SPGW service
- **Deployment Environment:** Vagrant (AGW)

### Setup Details
The S1APTester ([https://github.com/facebookexperimental/S1APTester](https://github.com/facebookexperimental/S1APTester)) was used along with Magma EPC in Vagrant based setup to reproduce the issue.

### Describe the Issue
The Non-Sanity test case **s1aptests/test_paging_with_mme_restart.py** is failing randomly for some iterations (usually in 2nd iteration only) with assertion error. The test case is expecting the initial context setup indication message after service request but receives the retransmitted paging message. This leads to assertion error.
[test_paging_with_mme_restart.py.zip](https://github.com/magma/magma/files/7298071/test_paging_with_mme_restart.py.zip)



The execution log from S1APTester terminal as follows:
```
. /home/vagrant/build/python/bin/activate
echo "Running test: s1aptests/test_paging_with_mme_restart.py"
Running test: s1aptests/test_paging_with_mme_restart.py
timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_paging_with_mme_restart.xml -x -s s1aptests/test_paging_with_mme_restart.py || exit 1
Start time 01:27:28
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config

**********************TPT open server******** 0

 Server Socket FD =[13]
************* Testing *Time elapsed(in usec): for timer expiry               :8287 usec
AGW is stateless
APN Correction configured
Health service is disabled
Using subscriber IMSI IMSI001010000000001
Using IMEI 3805468432113171
************************* UE device config for ue_id  1
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* Running End to End attach for UE id  1
ue id 1 found

*************NbUiNbuHdlUeRadioCapMsg: Send UE cap msg SUCCESS*********
PDN_TYPE: 1
IPv4 PDN_ADDRESS:192 168 128 12
************************* Sending UE context release request  for UE id  1
Deleting Ue Context from Traffic Handler
************************* Running UE downlink (UDP) for UE id  1
-----------------------------------------------------------
Server listening on 7001
-----------------------------------------------------------
************************ Received Paging Indication
************************* Restarting MME service on gateway
Waiting for 0 seconds
Waiting for 1 seconds
Waiting for 2 seconds
Waiting for 3 seconds
Waiting for 4 seconds
Waiting for 5 seconds
Waiting for 6 seconds
Waiting for 7 seconds
Waiting for 8 seconds
Waiting for 9 seconds
Waiting for 10 seconds
Waiting for 11 seconds
Waiting for 12 seconds
Waiting for 13 seconds
Waiting for 14 seconds
Waiting for 15 seconds
Waiting for 16 seconds
Waiting for 17 seconds
Waiting for 18 seconds
Waiting for 19 seconds
Waiting for 20 seconds
Waiting for 21 seconds
Waiting for 22 seconds
Waiting for 23 seconds
Waiting for 24 seconds
Waiting for 25 seconds
Waiting for 26 seconds
Waiting for 27 seconds
Waiting for 28 seconds
Waiting for 29 seconds
Accepted connection from 192.168.129.42, port 41035
[ 30] local 192.168.128.12 port 7001 connected to 192.168.129.42 port 47673
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[ 30]   0.00-1.00   sec  0.00 Bytes  0.00 bits/sec  0.000 ms  0/0 (0%)
[ 30]   1.00-1.16   sec  0.00 Bytes  0.00 bits/sec  0.000 ms  0/0 (0%)
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[ 30]   0.00-1.16   sec  0.00 Bytes  0.00 bits/sec  0.000 ms  0/0 (0%)
************************* send SCTP SHUTDOWN
Deleting Ue Context from Traffic Handler
Keys left in Redis (list should be empty)[
 IMSI001010000000001:MME
IMSI001010000000001:SPGW
IMSI001010000000001.magma.ipv4,ipv4:mobilityd_ipdesc_record

nb_ue_attached: 1
]
Entries in s1ap_imsi_map (should be zero): 1
Entries left in hashtables (should be zero): 3
Entries in mme_ueip_imsi_map (should be zero): 1
F
======================================================================
FAIL: The test case validates resumption of Paging Response Timer
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/test_paging_with_mme_restart.py", line 137, in test_paging_with_mme_restart
    response.msg_type, s1ap_types.tfwCmd.INT_CTX_SETUP_IND.value,
AssertionError: 40 != 53
-------------------- >> begin captured logging << --------------------
root: ERROR: Could not load magmad yml config for mconfig modules
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 17
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 17
root: INFO: Adding subscriber : IMSI001010000000001
root: DEBUG: s1ap response expected, received: 12, 12
root: DEBUG: s1ap message expected, received: 68, 68
root: INFO: Deleting subscriber : IMSI001010000000001
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 1 test in 65.196s
```


This is an extension of the issue #5842

**To Reproduce**
1) Run the test case individually "make integ_test TESTS=s1aptests/test_paging_with_mme_restart.py"

--- a/lte/gateway/python/integ_tests/defs.mk
+++ b/lte/gateway/python/integ_tests/defs.mk
@@ -160,6 +160,7 @@ s1aptests/test_guti_attach_with_zero_mtmsi.py \
 s1aptests/test_restore_mme_config_after_sanity.py
 
 NON_SANITY_TESTS = s1aptests/test_modify_config_for_non_sanity.py \
+s1aptests/test_paging_with_mme_restart.py \
 s1aptests/test_no_identity_rsp_with_mme_restart.py \
 s1aptests/test_agw_offload_idle_active_ue.py \
 s1aptests/test_standalone_pdn_conn_req_with_apn_correction.py \
@@ -220,7 +221,6 @@ s1aptests/test_restore_config_after_non_sanity.py
 
 # Non-Sanity: Flaky Test Cases
 # s1aptests/test_attach_detach_two_pdns_with_tcptraffic.py \ GitHubIssue 9670
-# s1aptests/test_paging_with_mme_restart.py \ GitHubIssue 9519
 # s1aptests/test_agw_offload_mixed_idle_active_multiue.py \ GitHubIssue 6063
 # s1aptests/test_attach_ul_udp_data_multi_ue.py \ Fails randomly with connection refused
 # s1aptests/test_attach_dl_udp_data_multi_ue.py \ Fails randomly with connection refused

--- a/lte/gateway/python/integ_tests/s1aptests/test_paging_with_mme_restart.py
+++ b/lte/gateway/python/integ_tests/s1aptests/test_paging_with_mme_restart.py
@@ -21,12 +21,16 @@ from s1ap_utils import MagmadUtil
 
 
 class TestPagingWithMmeRestart(unittest.TestCase):
+    """Unittest: TestPagingWithMmeRestart"""
+
     def setUp(self):
+        """Initialize before test case execution"""
         self._s1ap_wrapper = s1ap_wrapper.TestWrapper(
             stateless_mode=MagmadUtil.stateless_cmds.ENABLE,
         )
 
     def tearDown(self):
+        """Cleanup after test case execution"""
         self._s1ap_wrapper.cleanup()
 
     def test_paging_with_mme_restart(self):
@@ -45,7 +49,6 @@ class TestPagingWithMmeRestart(unittest.TestCase):
         Step6 : Expecting normal flow of DL data
         """
         self._s1ap_wrapper.configUEDevice(1)
-        time.sleep(20)
         req = self._s1ap_wrapper.ue_req
         ue_id = req.ue_id
         print(
@@ -79,11 +82,13 @@ class TestPagingWithMmeRestart(unittest.TestCase):
             gpp_types.CauseRadioNetwork.USER_INACTIVITY.value
         )
         self._s1ap_wrapper.s1_util.issue_cmd(
-            s1ap_types.tfwCmd.UE_CNTXT_REL_REQUEST, ue_cntxt_rel_req,
+            s1ap_types.tfwCmd.UE_CNTXT_REL_REQUEST,
+            ue_cntxt_rel_req,
         )
         response = self._s1ap_wrapper.s1_util.get_response()
         self.assertEqual(
-            response.msg_type, s1ap_types.tfwCmd.UE_CTX_REL_IND.value,
+            response.msg_type,
+            s1ap_types.tfwCmd.UE_CTX_REL_IND.value,
         )
 
         time.sleep(0.3)
@@ -92,7 +97,9 @@ class TestPagingWithMmeRestart(unittest.TestCase):
             ue_id,
         )
         with self._s1ap_wrapper.configDownlinkTest(
-            req, duration=1, is_udp=True,
+            req,
+            duration=1,
+            is_udp=True,
         ) as test:
             response = self._s1ap_wrapper.s1_util.get_response()
             self.assertTrue(response, s1ap_types.tfwCmd.UE_PAGING_IND.value)
@@ -108,21 +115,6 @@ class TestPagingWithMmeRestart(unittest.TestCase):
                 print("Waiting for", j, "seconds")
                 time.sleep(1)
 
-            ''' Note: Below commented lines needs to be uncommented in case if
-             Paging Response timer is configured more than mme restart time
-             (~40 seconds) because on expiry MME re-transmits the Paging
-             Indication again
-
-             Currently Paging Response timer is set to 4 seconds defined by
-             macro, MME_APP_PAGING_RESPONSE_TIMER_VALUE in mme_app_ue_context.h
-             file With current timer value, Paging Indication is not
-             re-transmitted because timer expires before MME restarts '''
-
-            '''print("************************ wait on Paging Indication");
-            response = self._s1ap_wrapper.s1_util.get_response()
-            self.assertTrue(response, s1ap_types.tfwCmd.UE_PAGING_IND.value)
-            print("************************ Received Paging Indication");'''
-
             # Send service request to reconnect UE
             ser_req = s1ap_types.ueserviceReq_t()
             ser_req.ue_Id = ue_id
@@ -130,20 +122,42 @@ class TestPagingWithMmeRestart(unittest.TestCase):
             ser_req.ueMtmsi.pres = False
             ser_req.rrcCause = s1ap_types.Rrc_Cause.TFW_MT_ACCESS.value
             self._s1ap_wrapper.s1_util.issue_cmd(
-                s1ap_types.tfwCmd.UE_SERVICE_REQUEST, ser_req,
+                s1ap_types.tfwCmd.UE_SERVICE_REQUEST,
+                ser_req,
             )
-            response = self._s1ap_wrapper.s1_util.get_response()
+
+            # It has been observed that despite getting the restart command on
+            # time, MME sometimes restarts after a delay of 5-6 seconds.
+            # Currently Paging response timer is set to 4 seconds defined by
+            # macro, MME_APP_PAGING_RESPONSE_TIMER_VALUE in the file
+            # mme_app_ue_context.h. If MME restarts after expiry of paging
+            # response timer of 4 sec, it will again send the Paging indication
+            resp_count = 0
+            while True:
+                response = self._s1ap_wrapper.s1_util.get_response()
+
+                if response.msg_type == s1ap_types.tfwCmd.UE_PAGING_IND.value:
+                    resp_count += 1
+                    print(
+                        "******************** Ignoring re-transmitted (",
+                        resp_count,
+                        ") Paging indication",
+                    )
+                else:
+                    break
+
             self.assertEqual(
-                response.msg_type, s1ap_types.tfwCmd.INT_CTX_SETUP_IND.value,
+                response.msg_type,
+                s1ap_types.tfwCmd.INT_CTX_SETUP_IND.value,
             )
             test.verify()
 
         time.sleep(0.5)
         # Now detach the UE
         self._s1ap_wrapper.s1_util.detach(
-            ue_id, s1ap_types.ueDetachType_t.UE_NORMAL_DETACH.value, True,
+            ue_id,
+            s1ap_types.ueDetachType_t.UE_NORMAL_DETACH.value,
         )
-        time.sleep(0.5)
 
 
 if __name__ == "__main__":

